pkg/edit/editor_test.go

```diff
@@ -37,6 +37,37 @@ func TestEditor_DoesCommandWithLeadingSpaceToHistory(t *testing.T) {
 	testCommands(t, f.Store /* no commands */)
 }
 
+func TestEditor_FiltersCommandBeforeAddingToHistory(t *testing.T) {
+	noHistory := "_EMPTY_"
+
+	test := func(c, input, exp string) {
+		f := setup(rc(c))
+		defer f.Cleanup()
+
+		feedInput(f.TTYCtrl, input)
+		f.Wait()
+
+		if exp == noHistory {
+			testCommands(t, f.Store)
+		} else {
+			testCommands(t, f.Store, exp)
+		}
+	}
+
+	// item is not a callable, invalid
+	test("edit:add-cmd-filters = [$false]", "echo\n", "echo")
+	// function has no output, invalid
+	test("edit:add-cmd-filters = [[_]{}]", "echo\n", "echo")
+	test("edit:add-cmd-filters = [[_]{ put $true }]", "echo\n", "echo")
+	test("edit:add-cmd-filters = [[_]{ put $false }]", "echo\n", noHistory)
+	// filters are concatenated internally with logical "and"
+	test("edit:add-cmd-filters = [[_]{ put $false } [_]{ put $true }]", "echo\n", noHistory)
+	test("edit:add-cmd-filters = [[_]{ put $true } [_]{ put $false }]", "echo\n", noHistory)
+
+	test("edit:add-cmd-filters = [[cmd]{ ==s $cmd echo }]", "echo\n", "echo")
+	test("edit:add-cmd-filters = [[cmd]{ ==s $cmd echo }]", "echo x\n", noHistory)
```

Comment:
Prefer one of the following two styles:

* Put all test cases in a table, and iterate through the table in this test. If you do that, add a `name` field to the table, and use [`t.Run`](https://godoc.org/testing#T.Run) so that each case is run as a subtest.

* Make each of the test a top-level `TestBeforeAddCmd_Foo` function; also make the `test` closure a top-level, unexported `testBeforeAddCmd` function, so that all the `TestBeforeAddCmd_Foo` function just calls `testBeforeAddCmd`.